File: src/com/easy/SubtreeOfAnotherTree_572.java
/**
 * 
 */
package com.easy;


/**
 * @author Chris Lin
 * @version 1.0
 */
public class SubtreeOfAnotherTree_572 {
    
    // Recursive version
    // Time Complexity O(N^2), m for numbers of tree and n for numbers of subtree
    // Runtime: 5 ms, faster than 98.89%
    // Memory Usage: 41.3 MB, less than 88.99%
    public boolean isSubtree(TreeNode s, TreeNode t) {
        if(s == null) {
            return false;
        }
        return isSubTreeWithRoot(s, t) || isSubtree(s.left, t) || isSubtree(s.right, t);
    }

    private boolean isSubTreeWithRoot(TreeNode s, TreeNode t) {
        if (s == null && t == null) {
            return true;
        }
        if (s == null || t == null) {
            return false;
        }
        if (s.val != t.val) {
            return false;
        }
        return isSubTreeWithRoot(s.left, t.left) && isSubTreeWithRoot(s.right, t.right);
    }
    
    // use inorder traversal and judge String for substring or not
    // Time Complexity O(N)
    // Runtime: 10 ms, faster than 64.02%
    // Memory Usage: 45.1 MB, less than 5.41%
    public boolean isSubtree2(TreeNode s, TreeNode t) {
        String tree = preorder(s);
        String subTree = preorder(t);
        return (tree.indexOf(subTree) >= 0);
    }
    
    private String preorder(TreeNode mNode) {
        if(mNode == null) {
            return "#";
        }
        return "*" + mNode.val + " " + preorder(mNode.left) + " " + preorder(mNode.right);
    }
}

File: Golang_src/easy/subtreeOfAnotherTree_572.go
package easy

import (
	"strconv"
	"strings"
)

// TreeNode for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Use node comparison:
// Time Complexity: O(m*n)
// Space Complexity:O(n)
// Runtime: 12 ms, faster than 99.07%
// Memory Usage: 6.2 MB, less than 10.19%
func isSubtree(s *TreeNode, t *TreeNode) bool {
	if s == nil {
		return false
	}

	return treeCompare(s, t) || isSubtree(s.Left, t) || isSubtree(s.Right, t)
}

func treeCompare(p *TreeNode, q *TreeNode) bool {
	if p == nil && q == nil {
		return true
	}
	if p == nil || q == nil {
		return false
	}
	if p.Val != q.Val {
		return false
	}

	return treeCompare(p.Left, q.Left) && treeCompare(p.Right, q.Right)
}

// Use Preorder Traversal :
// Time Complexity: O(m*n)
// Space Complexity:O(max(m,n))
// Runtime: 16 ms, faster than 93.98%
// Memory Usage: 7 MB, less than 10.19%
func isSubtree2(s *TreeNode, t *TreeNode) bool {
	strS := preorder(s)
	strT := preorder(t)

	return strings.Contains(strS, strT)
}

func preorder(root *TreeNode) string {
	if root == nil {
		return "*"
	}

	return "#" + strconv.Itoa(root.Val) + preorder(root.Left) + preorder(root.Right)
}
